packages/block-library/src/image/edit.js

```diff
@@ -236,27 +284,129 @@ class ImageEdit extends Component {
+			return;
+		}
+
+		const height = round( fileHeight * ( width / fileWidth ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
+	}
+
+	updateHeight( height, fileWidth, fileHeight, userSet ) {
+		height = parseInt( height, 10 );
+
+		// Reset the image size when the user deletes the value.
+		if ( ! height || ! fileWidth || ! fileHeight ) {
+			this.resetWidthHeight();
+			return;
+		}
+
+		const width = round( fileWidth * ( height / fileHeight ) );
+		this.setWidthHeight( width, height, fileWidth, fileHeight, userSet );
 	}
 
-	updateImageURL( url ) {
-		this.props.setAttributes( { url, width: undefined, height: undefined } );
+	setWidthHeight( width, height, fileWidth, fileHeight, userSet ) {
+		// Set image size and also whether set directly by the user.
+		userSet = userSet || undefined;
+
+		this.props.setAttributes( {
+			width,
+			height,
+			fileWidth,
+			fileHeight,
+			userSet,
+		} );
 	}
 
-	updateWidth( width ) {
-		this.props.setAttributes( { width: parseInt( width, 10 ) } );
+	resetWidthHeight( fileWidth, fileHeight ) {
+		if ( fileWidth && fileHeight ) {
+			this.props.setAttributes( {
+				fileWidth,
+				fileHeight,
+			} );
+		}
+
+		this.props.setAttributes( {
+			width: undefined,
+			height: undefined,
+			userSet: undefined,
+		} );
 	}
 
-	updateHeight( height ) {
-		this.props.setAttributes( { height: parseInt( height, 10 ) } );
+	/**
+	 * Helper function to test if aspect ratios for two images match.
+	 *
+	 * @param {number} fullWidth  Width of the image in pixels.
+	 * @param {number} fullHeight Height of the image in pixels.
+	 * @param {number} targetWidth  Width of the smaller image in pixels.
+	 * @param {number} targetHeight Height of the smaller image in pixels.
+	 * @return {boolean} True if aspect ratios match within 1px. False if not.
+	 */
+	imageMatchesRatio( fullWidth, fullHeight, targetWidth, targetHeight ) {
+		if ( ! fullWidth || ! fullHeight || ! targetWidth || ! targetHeight ) {
+			return false;
+		}
+
+		const { width, height } = this.constrainImageDimensions( fullWidth, fullHeight, targetWidth );
+
+		// If the image dimensions are within 1px of the expected size, we consider it a match.
+		return ( Math.abs( width - targetWidth ) <= 1 && Math.abs( height - targetHeight ) <= 1 );
 	}
 
-	updateDimensions( width = undefined, height = undefined ) {
-		return () => {
-			this.props.setAttributes( { width, height } );
+	constrainImageDimensions( fullWidth, fullHeight, targetWidth ) {
+		const ratio = targetWidth / fullWidth;
+
+		// Very small dimensions may result in 0, 1 should be the minimum.
+		const height = Math.max( 1, round( fullHeight * ratio ) );
+		let width = Math.max( 1, round( fullWidth * ratio ) );
+
+		// Sometimes, due to rounding, we'll end up with a result like this: 465x700 in a 177x177 box is 117x176... a pixel short.
+		if ( width === targetWidth - 1 ) {
+			width = targetWidth; // Round it up
+		}
+
+		return {
+			width: width,
+			height: height,
```

Comment:
`constrainImageDimensions` is pretty much a copy of core's `wp_constrain_dimensions()`: https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L400. We can change/improve it as long as it works in exactly the same way (and is "easy to read") :)

> since this doesn't depend on any instance variables, might as well put it outside of the class and make it a simple helper function

Sure, it probably can be re-used somewhere else.